src/backend/distributed/commands/function.c

```diff
@@ -91,6 +156,305 @@ CreateFunctionDDLCommandsIdempotent(const ObjectAddress *functionAddress)
+							   functionName, sourceRelationName),
+						errdetail("Citus currently only supports colocating function "
+								  "with distributed tables that are created using "
+								  "streaming replication model."),
+						errhint("When distributing tables make sure that "
+								"\"citus.replication_model\" is set to \"streaming\"")));
+	}
+
+	sourceDistributionColumnType = sourceDistributionColumn->vartype;
+	if (sourceDistributionColumnType != distributionColumnType)
+	{
+		char *functionName = get_func_name(functionOid);
+		char *sourceRelationName = get_rel_name(sourceRelationId);
+
+		ereport(ERROR, (errmsg("cannot colocate function \"%s\" and table \"%s\" "
+							   "because distribution column types don't match",
+							   sourceRelationName, functionName)));
+	}
+}
+
+
+/*
+ * UpdateFunctionDistributionInfo gets object address of a function and
+ * updates its distribution_argument_index and colocationId in pg_dist_object.
+ */
+static void
+UpdateFunctionDistributionInfo(const ObjectAddress *distAddress,
+							   int distribution_argument_index,
+							   int colocationId)
+{
+	int paramCount = 5;
+	Oid paramTypes[5] = {
+		INT4OID,
+		INT4OID,
+		OIDOID,
+		OIDOID,
+		INT4OID
+	};
+	Datum paramValues[5] = {
+		Int32GetDatum(distribution_argument_index),
+		Int32GetDatum(colocationId),
+		ObjectIdGetDatum(distAddress->classId),
+		ObjectIdGetDatum(distAddress->objectId),
+		Int32GetDatum(distAddress->objectSubId)
+	};
+	int spiStatus = 0;
+
+	char *updateQuery =
+		"UPDATE citus.pg_dist_object SET distribution_argument_index = $1, "
+		"colocationid = $2 WHERE classid = $3 AND objid=$4 AND objsubid=$5";
+
+	/* UPDATE query can only be useful if the entry is already in the pg_dist_object */
+	Assert(IsObjectDistributed(distAddress));
+
+	spiStatus = ExecuteCommandAsSuperuser(updateQuery, paramCount, paramTypes,
+										  paramValues);
+	if (spiStatus < 0)
+	{
+		ereport(ERROR, (errmsg("failed to insert object into citus.pg_dist_object")));
+	}
+}
+
+
+/*
+ * ResetFunctionDistributionInfo gets object address of a function and
+ * updates its distribution_argument_index and colocationId in pg_dist_object
+ * to NULL.
+ */
+static void
+ResetFunctionDistributionInfo(const ObjectAddress *distAddress)
+{
+	int paramCount = 3;
+	Oid paramTypes[3] = {
+		OIDOID,
+		OIDOID,
+		INT4OID
+	};
+	Datum paramValues[3] = {
+		ObjectIdGetDatum(distAddress->classId),
+		ObjectIdGetDatum(distAddress->objectId),
+		Int32GetDatum(distAddress->objectSubId)
+	};
+	int spiStatus = 0;
+
+	char *updateQuery =
+		"UPDATE citus.pg_dist_object SET distribution_argument_index = NULL, "
+		"colocationid = NULL WHERE classid = $1 AND objid=$2 AND objsubid=$3";
+
+	/* UPDATE query can only be useful if the entry is already in the pg_dist_object */
+	Assert(IsObjectDistributed(distAddress));
+
+	spiStatus = ExecuteCommandAsSuperuser(updateQuery, paramCount, paramTypes,
+										  paramValues);
```

Comment:
We probably should not overuse this pattern. In general it is a bit dangerous to enter SQL land as superuser. If an administrator sets up a trigger on citus.pg_dist_object then a user might be able to exploit the fact that the trigger is called as superuser. I don't have any specific concerns here, but I'm weary of dangerous patterns being repeated later on. We should be adding some warning signs.

Reply:
Good to know about this case. I used it just for simplicity (with copy & paste), now switched to `heap_tuple_modify()`

Reply:
Also added comment to `ExecuteCommandAsSuperuser `